### PR TITLE
feat: add support for the strategy parameter in component-name-unique rule

### PR DESCRIPTION
**Pull request by harshit078**
## Reference

https://github.com/Redocly/redocly-cli/issues/2898

## Check yourself

- [x] This PR follows the [contributing guide](https://github.com/Redocly/redocly-cli/blob/main/CONTRIBUTING.md#pull-request-guidelines) 
- [x] All new/updated code is covered by tests
- [ ] Core code changed? - Tested with other Redocly products (internal contributions only)
- [ ] New package installed? - Tested in different environments (browser/node)
- [x] Documentation update has been considered 

## Security

- [x] The security impact of the change has been considered 
- [x] Code follows company security practices and guidelines

---

> [!NOTE]
> **Medium Risk**
> Lint and bundle naming logic now share utilities; incorrect title-strategy edge cases could cause false positives/negatives before bundling, but behavior is heavily covered by new tests.
> 
> **Overview**
> Adds a **`strategy`** option (`basename` | `title`, default `basename`) to the **`component-name-unique`** lint rule so uniqueness checks mirror **`bundle`**’s `--component-names-strategy`. With **`title`**, external schemas are keyed by sanitized PascalCase **`title`** (using the composed **`$ref`** target when present); root **`components.schemas`** keys stay as authored. Missing **`title`** on schemas that bundle would rename is reported (with **`schemas`** severity when set), and uniqueness still falls back to file names for those cases.
> 
> Shared helpers **`componentNameFromTitle`** and **`effectiveRefTarget`** are extracted and wired into **`bundle-visitor`** so naming and ref-chain behavior stay aligned. Docs and a changeset describe configuring **`strategy`** alongside bundle.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8142325da993b362f64056ae31d9588f4e68bfde. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by adamaltman** [comment]
Interesting idea!

File: .changeset/vast-kids-add.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@redocly/cli': minor
+'@redocly/respect-core': minor
```
Comment (kanoru3101) [inline .changeset/vast-kids-add.md]:
```suggestion
'@redocly/cli': minor
'@redocly/openapi-core': minor
```

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 2]
```diff
@@ -147,6 +160,23 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
...
   }
+
+  function getTitleComponentName(
+    typeName: string,
+    resolved: { node: unknown; location: Location }
+  ): string | null {
+    if (
+      !useTitleStrategy ||
+      typeName !== TYPE_NAME_SCHEMA ||
+      resolved.location.source.absoluteRef === rootSourceRef
```
Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts]: The rule skips the title strategy for every schema in the root file. The `bundle` command does not skip those schemas. It renames a root schema if an external file refers to it. So two schemas end up wanting the same name and the rule stays quiet.
Example
```
# openapi.yaml
components:
  schemas:
    Foo:
      title: Bar thing
      type: object
# ...a response in this file refers to ./Other.yaml
```
```
# Other.yaml
title: Bar thing
type: object
properties:
  inner:
    $ref: './openapi.yaml#/components/schemas/Foo'
```
The bundle is still written, but `Other.yaml` gets `Other`, not the name from its title. The rule sees no problem here, the bundler does both and should see the same. Please check where the $ref comes from, not only where it points.

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 3]
```diff
@@ -147,6 +160,23 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
...
+    ) {
+      return null;
+    }
+
+    const { node } = resolved;
+    const title = isPlainObject(node) && isString(node.title) ? node.title.trim() : '';
+    return title === '' ? null : componentNameFromTitle(title);
```
Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts]: If a schema has no `title`, the rule uses the file name and says nothing. The bundler doesn't stop.

```
# openapi.yaml
openapi: 3.0.0
info:
  title: Referenced schema without a title
  version: 1.0.0
paths:
  /carts:
    get:
      responses:
        '200':
          description: ok
          content:
            application/json:
              schema:
                $ref: './Cart.yaml'
```
```
# Cart.yaml - no title here
type: object
properties:
  total:
    type: number
```

Lint says the description is fine, and then there is no bundle at all. The two commands disagree again, and this is the most common way the title strategy breaks, so the rule should report it:

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 4]
```diff
@@ -147,6 +160,23 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
...
+    if (
+      !useTitleStrategy ||
+      typeName !== TYPE_NAME_SCHEMA ||
+      resolved.location.source.absoluteRef === rootSourceRef
+    ) {
+      return null;
+    }
+
+    const { node } = resolved;
+    const title = isPlainObject(node) && isString(node.title) ? node.title.trim() : '';
```
Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts]: This duplicates code that already exists; you should consider reusing the existing logic

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 5]
```diff
@@ -55,11 +60,19 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
           const resolvedRef = resolve(ref);
           if (!resolvedRef.location) return;
 
-          addComponentFromAbsoluteLocation(typeName, resolvedRef.location);
+          const titleName = getTitleComponentName(typeName, resolvedRef);
+          if (titleName) {
+            addFoundComponent(typeName, titleName, resolvedRef.location);
```
Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts]: The report points at the start of the file, but the line to change is the title. Pass `resolvedRef.location.child('title')` here, in the same place the bundler reports at `bundle-visitor.ts`.

**Review by kanoru3101 — COMMENTED** [review]
Thank you for your contribution.

Three descriptions now behave differently under `lint` and `bundle`: two in my comments, plus the Bugbot one above — that one is correct, I reproduced it.
Left to do:
* make the rule agree with the bundler in all three cases;
* mirror the bundler's tests  — none of the three cases is covered today.
The rest is in the inline comments.

**Comment by kanoru3101** [comment]
Hey @harshit078 !
Just checking in on this PR. If you’re not planning to continue with it, I’d be happy to take it over and keep working on it. Let me know if that works for you!

**Comment by harshit078** [comment]
Hey @kanoru3101 , I completely forgot to push my fixes for this PR and left it stale. I'll push the solution for it. Thanks !

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 6]
```diff
@@ -55,11 +60,19 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
...
+          } else {
+            addComponentFromAbsoluteLocation(typeName, resolvedRef.location);
+          }
         }
       },
     },
     Root: {
+      enter(_: AnyOas3Definition, { location }: UserContext) {
+        rootSourceRef = location.source.absoluteRef;
+      },
```
[the remark this answers is not in the document]
Comment (kanoru3101) [reply packages/core/src/rules/oas3/component-name-unique.ts:83]: Reopening this. it still reproduces on the current head, the fix commit addressed a different problem.
```yaml
# openapi.yaml
components:
  schemas:
    Order:
      type: object
    OrderRef:
      $ref: './Order.yaml'
```
```
# Order.yaml
title: Order model
type: object
```
With `strategy: title`, the `lint` command reports `Component 'schemas/Order' is not unique` twice. `bundle --component-names-strategy=title` is clean — the external schema becomes `OrderModel` and never competes for the name `Order`.

The fix is in the `NamedSchemas` visitor skip the basename registration when the title strategy owns the schema; all the current rule tests pass with this:
```js
Schema(_: Oas3Schema | Oas3_1Schema, { location }: UserContext) {
  if (useTitleStrategy && location.source.absoluteRef !== rootSourceRef) return;
  addComponentFromAbsoluteLocation(TYPE_NAME_SCHEMA, location);
},
```

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 7]
```diff
@@ -49,26 +53,46 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
...
 
+          if (usesTitleStrategy(typeName, location, resolvedRef.location)) {
+            const { title, name } = componentNameFromTitle(resolvedRef.node);
+            if (title) {
+              addFoundComponent(
+                typeName,
+                name,
+                resolvedRef.location,
+                resolvedRef.location.child('title')
+              );
```
[the remark this answers is not in the document]
Comment (kanoru3101) [reply packages/core/src/rules/oas3/component-name-unique.ts]: This one also still reproduces. Example:
```yaml
# Mid.yaml
$ref: './Final.yaml'
title: Mid title
```
```yaml
# Final.yaml
title: Final title
type: object
```
```yaml
# Other2.yaml
title: Mid title
type: object
```
Root refers to `./Mid.yaml` and `./Other2.yaml`: `lint` is quiet, `bundle` warns `Title "Mid title" maps to component name MidTitle, already used by another schema`.

File: packages/core/src/rules/oas3/__tests__/component-name-unique.test.ts [hunk 8]
```diff
@@ -986,4 +986,413 @@ describe('Oas3 component-name-unique', () => {
       `);
     });
   });
+
+  describe('strategy: title', () => {
```
Comment (kanoru3101) [inline packages/core/src/rules/oas3/__tests__/component-name-unique.test.ts:990]: Please add tests for the re-export case above, for a title that needs sanitizing (`Order/model`), and for non-schema components staying untouched. The bundler has exactly this set at `bundle.test.ts:1104`. The rule should mirror it.

**Review by kanoru3101 — COMMENTED** [review]
The previous issues have been resolved; thank you. However, two of the resolved issues on Bugbot still allow the error to be reproduced. I have reopened both of them with reproduction examples.

File: packages/core/src/utils/effective-ref-target.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { type Location } from '../ref-utils.js';
+import { type ResolvedRefChainHop } from '../resolve.js';
+
+export type RefTarget = { node: unknown; location: Location };
+
+export function effectiveRefTarget(resolved: {
```
Comment (kanoru3101) [inline packages/core/src/utils/effective-ref-target.ts:7]: The explanatory comment from the old `effectiveTarget` got lost in the move: "a composed $ref in the chain is the effective target, so the composition survives bundling". Please put it back on this function — without it, `chain?.[0]` doesn't explain why the first hop wins.

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 10]
```diff
@@ -83,13 +102,30 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
...
         });
+
+        for (const location of schemasWithoutTitle.values()) {
+          const problem: Problem = {
+            message:
+              'Schema must define a `title` when using `strategy: title`. Bundling fails without it.',
+            location,
+            reference: 'https://redocly.com/docs/cli/rules/oas/component-name-unique',
+          };
+          if (options.schemas) {
```
Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts:115]: This branch has no test: with `schemas: warn` the missing-title report should come out as a warning. One test covers it — `{ severity: 'error', schemas: 'warn', strategy: 'title' }` plus a referenced schema without a title, asserting the problem has `"severity": "warn"` in the snapshot. Something like `should use the schemas severity for a missing title report`.

File: docs/@v2/rules/oas/component-name-unique.md [hunk 11]
```diff
@@ -48,8 +49,26 @@ rules:
     parameters: off
     responses: warn
     requestBodies: warn
+    strategy: basename
 ```
 
+### Component names strategy
+
+The rule predicts the component names that `bundle` produces, so `strategy` must match the
+[`--component-names-strategy`](../../commands/bundle.md#configure-the-component-names-strategy) option you bundle with.
```
Comment (kanoru3101) [inline docs/@v2/rules/oas/component-name-unique.md]:
```suggestion
The rule predicts the component names that `bundle` produces, so `strategy` must match the [`--component-names-strategy`](../../commands/bundle.md#configure-the-component-names-strategy) option you bundle with.
```

File: docs/@v2/rules/oas/component-name-unique.md [hunk 12]
```diff
@@ -48,8 +49,26 @@ rules:
...
+
+With the default `basename`, a schema pulled in from another file is named after the `$ref` fragment or the file name.
+Two files both called `Order.yaml` therefore collide, and the rule reports them.
+
+With `title`, the same schemas are named after their `title` field instead.
+Two files called `Order.yaml` with the titles `Order model` and `Order request` become `OrderModel` and `OrderRequest`, so the rule no longer reports them.
+Two schemas in differently named files that share a title do collide, and the rule reports those instead.
+
+The `title` strategy applies to every schema that `bundle` renames, which is every schema reached by a `$ref` that crosses a file boundary.
```
Comment (kanoru3101) [inline docs/@v2/rules/oas/component-name-unique.md]:
```suggestion
The `title` strategy applies to every schema that `bundle` renames every `$ref` except those fully inside the root document.
```

**Review by kanoru3101 — COMMENTED** [review]
Left a few minor comments

File: packages/core/src/utils/effective-ref-target.ts (hunk 9 above)
Comment (harshit078) [reply packages/core/src/utils/effective-ref-target.ts:7]: Hi, Added the comment back !

File: packages/core/src/rules/oas3/component-name-unique.ts (hunk 10 above)
Comment (harshit078) [reply packages/core/src/rules/oas3/component-name-unique.ts:115]: Agreed, I have added the required test.

File: docs/@v2/rules/oas/component-name-unique.md [hunk 13]
```diff
@@ -48,8 +49,25 @@ rules:
...
+The rule predicts the component names that `bundle` produces, so `strategy` must match the [`--component-names-strategy`](../../commands/bundle.md#configure-the-component-names-strategy) option you bundle with.
+
+With the default `basename`, a schema pulled in from another file is named after the `$ref` fragment or the file name.
+Two files both called `Order.yaml` therefore collide, and the rule reports them.
+
+With `title`, the same schemas are named after their `title` field instead.
+Two files called `Order.yaml` with the titles `Order model` and `Order request` become `OrderModel` and `OrderRequest`, so the rule no longer reports them.
+Two schemas in differently named files that share a title do collide, and the rule reports those instead.
+
+The `title` strategy applies to every schema that `bundle` renames every `$ref` except those fully inside the root document.
```
Comment (kanoru3101) [inline docs/@v2/rules/oas/component-name-unique.md]: The dash got lost when the suggestion was applied — this now reads as "`bundle` renames every `$ref`".

```suggestion
The `title` strategy applies to every schema that `bundle` renames — every `$ref` except those fully inside the root document.
```

File: docs/@v2/commands/bundle.md [hunk 14]
```diff
@@ -205,3 +205,6 @@ All other characters, including non-ASCII letters such as `é` or `я`, are repl
 Schemas without `title` can't be named using the `--component-names-strategy=title` strategy.
 The bundling process reports an error for such schemas.
 {% /admonition %}
+
+To catch name collisions before bundling, set the matching `strategy` option on the
+[`component-name-unique`](../rules/oas/component-name-unique.md) rule.
```
Comment (kanoru3101) [inline docs/@v2/commands/bundle.md]: Semantic line breaks — this is one sentence, so one line:

```suggestion
To catch name collisions before bundling, set the matching `strategy` option on the [`component-name-unique`](../rules/oas/component-name-unique.md) rule.
```

File: packages/core/src/rules/oas3/component-name-unique.ts [hunk 15]
```diff
@@ -49,26 +54,40 @@ export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
...
 
+          const target = effectiveRefTarget(resolvedRef);
+
+          if (usesTitleStrategy(typeName, target.location, location)) {
+            const { title, name } = componentNameFromTitle(target.node);
+            if (title) {
+              addFoundComponent(typeName, name, target.location, target.location.child('title'));
+              return;
+            }
+            schemasWithoutTitle.set(target.location.absolutePointer.toString(), target.location);
```
Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts]: Take the fallback name from the composed hop, like the bundler does — otherwise the clash that `bundle` resolves as `Mid-2` (root `Mid` plus a composed titleless `Mid.yaml`) goes unreported. All the rule tests pass with this:

```suggestion
            schemasWithoutTitle.set(target.location.absolutePointer.toString(), target.location);
            addComponentFromAbsoluteLocation(typeName, target.location);
            return;
```

Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts]: One last thing, from the Bugbot thread above: when a composed `$ref` has no title, take the fallback name from the hop — that's what the bundler names it. Otherwise the clash it resolves as `Mid-2` (root `Mid` + composed titleless `Mid.yaml`) slips through lint. Tried it locally, all tests pass:

```suggestion
            schemasWithoutTitle.set(target.location.absolutePointer.toString(), target.location);
            addComponentFromAbsoluteLocation(typeName, target.location);
            return;
```

Worth a small test too: root `Mid` plus a composed titleless `Mid.yaml`.

Comment (kanoru3101) [inline packages/core/src/rules/oas3/component-name-unique.ts:71]: One last thing, from the Bugbot thread above: when a composed `$ref` has no title, take the fallback name from the hop — that's what the bundler names it. Otherwise the clash it resolves as `Mid-2` (root `Mid` + composed titleless `Mid.yaml`) slips through lint. Tried it locally, all tests pass:

```suggestion
            schemasWithoutTitle.set(target.location.absolutePointer.toString(), target.location);
            addComponentFromAbsoluteLocation(typeName, target.location);
            return;
```

File: .changeset/vast-kids-add.md [hunk 16]
```diff
@@ -0,0 +1,8 @@
+---
+'@redocly/cli': minor
+'@redocly/openapi-core': minor
+---
+
+Added a `strategy` option to the `component-name-unique` rule, matching the `--component-names-strategy` option of the `bundle` command.
+Set it to `title` to check the component names that bundling derives from each schema's `title`.
+With `strategy: title`, the rule also reports referenced schemas that have no `title`, because `bundle` can't name those and fails.
```
Comment (JLekawa) [inline .changeset/vast-kids-add.md]:
```suggestion
```

File: docs/@v2/rules/oas/component-name-unique.md [hunk 17]
```diff
@@ -48,8 +49,25 @@ rules:
...
+The rule predicts the component names that `bundle` produces, so `strategy` must match the [`--component-names-strategy`](../../commands/bundle.md#configure-the-component-names-strategy) option you bundle with.
+
+With the default `basename`, a schema pulled in from another file is named after the `$ref` fragment or the file name.
+Two files both called `Order.yaml` therefore collide, and the rule reports them.
+
+With `title`, the same schemas are named after their `title` field instead.
+Two files called `Order.yaml` with the titles `Order model` and `Order request` become `OrderModel` and `OrderRequest`, so the rule no longer reports them.
+Two schemas in differently named files that share a title do collide, and the rule reports those instead.
+
+The `title` strategy applies to every schema that `bundle` renames — every `$ref` except those fully inside the root document.
```
Comment (JLekawa) [inline docs/@v2/rules/oas/component-name-unique.md]:
```suggestion
The `title` strategy applies to every schema that `bundle` renames.
Every `$ref` except those fully inside the root document.
```

File: docs/@v2/rules/oas/component-name-unique.md [hunk 18]
```diff
@@ -48,8 +49,25 @@ rules:
...
+A referenced schema that has no `title` can't be named under this strategy, and `bundle` fails without producing a file.
+The rule reports those schemas so you find them before bundling.
```
Comment (JLekawa) [inline docs/@v2/rules/oas/component-name-unique.md]:
```suggestion
The rule reports these schemas, so you find them before bundling.
```

File: docs/@v2/rules/oas/component-name-unique.md [hunk 19]
```diff
@@ -48,8 +49,25 @@ rules:
...
+For the uniqueness check itself, such a schema still falls back to its file name, so a name collision is reported as well.
```
Comment (JLekawa) [inline docs/@v2/rules/oas/component-name-unique.md]:
```suggestion
For the uniqueness check itself, such schemas still fall back to their file names, so a name collision is reported as well.
```

**Review by JLekawa — APPROVED** [review]

**Review by kanoru3101 — APPROVED** [review]
Thank you for the contribution and for patiently working through all the review rounds - much appreciated!
